### ISSUE TITLE
[FEATURE]: allowing to prevent using `kind: Deploy``type: container` actions in production environments

### ISSUE THREAD

**Issue by stefreak** [garden-io/garden#5016]
### Background / Motivation

Using the `kind: Deploy` and `type: container` actions and other Garden features that use Kubernetes manifests that are bundled with the current Garden version, e.g. the default ingress controller, are great convenience features in some cases, but in production environments there are several good reasons not to use them:
- *Control*: Generates Kubernetes manifests, or manifests that ship with Garden, are not guaranteed to stay the same across Garden versions. Using the `kubernetes` deploy action gives you more control, even when the Garden version you use in CI changes.
- *GitOps*: The more about your infrastructure is being explicitly controlled through changes in your Git repository, the better. Explicitly declaring Kubernetes manifests not only allows you to use tools like ArgoCD and share manifests with Garden, but it also allows you to understand and auditing infrastructure changes by looking at the Git history.

### What should the user be able to do?

We should have an option that prevents deploying Kubernetes manifests that ship with Garden in production environments (`production: true`).
These include (not exhaustively):
- The ingress controller and the default backend
- Garden helper Pods, e.g Buildkit and the Util pods
- Using the deploy `type: container` 

By default we should consider warning the user, or asking interactively, before performing such operations.

In case of the `type: container` deploy action, we should offer an `eject` command that allows migration from `type: container` to `type: kubernetes` easily, and suggest doing so in the warning message in case it's being used in a production environment.

### Why do they want to do this? What problem does it solve?

Garden should encourage good practices in production environments.

### Suggested Implementation(s)

- Implement a flag that prevents using Kubernetes manifests that ship with Garden in production environments. This flag can be a provider-level configuration flag, e.g `enforceExplicitKubernetesManifests: "never"|"production-environments"|"all-environments"` in the kubernetes provider configuration.
- Make sure that in the entire code base that we fail before deploying manifests that have not been declared explicitly by the user in environments with `production: true`.
- Implement an `eject` command to help users migrate from the `container` deploy to the `kubernetes` deploy action
- By default, for backward compatibility, print a warning message (that can be disabled) when deploying manifests into production environments that have not been declared explicitly and suggest using the eject command.

### How important is this feature for you/your team?

🌵 Not having this feature makes using Garden painful

I chose this severity because in organisations that value *Control* and *GitOps*, this makes using Garden painful.

**Comment by twelvemo** [comment]
I agree that it would be good to think about ways we can integrate better with GitOps tools and make the prod flag more meaningful. However i am not sure we should enforce this behavior or introduce another flag, since it might be hard to understand. After some thought my preferred way would be another guide in our docs analogue to "Using Garden in CI" about "Using Garden in production". We could write down best-practices, provide an example and also point to limitations. When using the prod flag we could display a warning when either building containers and not fetching build images, using container deploys or using Gardens ingress controller and link to this guide.
I believe the ingress controller is disabled by default for the kubernetes provider already and only enabled by default for the local-kubernetes provider.

**Comment by stefreak** [comment]
I agree that the docs and printing a warning could be a first step.

Meanwhile I think that certain users will want to impose guardrails on Garden features (e.g. prohibiting the use of certain features that give away control). Garden should somehow facilitate that too.

**Comment by worldofgeese** [comment]
I love the `eject` command as one alternative to #4729. Another alternative would be to simply rebase `container` on our own fork of [The Helmet](https://github.com/companyinfo/helm-charts/tree/main/charts/helmet) Library Chart which would allow users to continue to treat `container` as an action of simplicity with greatly expanded access to keys.
